### PR TITLE
Dbplyr fix

### PR DESCRIPTION
**Pull request by ablack3**
Work in progress dplyr fix.

dplyr tests added and passing for redshift, postgres, sql server, oracle.

DBI::dbGetInfo needs to be implemented properly which requires getting the database version. This is important for sql server only so far.

### CODE REVIEW

**Comment by ablack3** [comment]
@schuemie - This seems to be getting closer. I did some deeper diving into DatabaseConnector and tried to separate the DBI focused refactoring from getting dplyr working. 

I'm not sure how to implement dbGetInfo for each connection type yet and added a temporary workaround that needs to be fixed. However postgresql, redshift, oracle, SQL server, and sqlite seem to be working with dplyr and I've added a simple test for each. That's probably the extent of what we can do without test databases for the other dbms.

**Comment by schuemie** [comment]
I'm not quite following how this works. How does dbplyr know that for PostgresSQL, `head(5)` needs to be translated to `LIMIT 5` for PostgresSQL, and to `TOP 5` for SQL Server? Or does it rely on the fetch size of the JDBC query function?

File: NAMESPACE [hunk 1]
```diff
@@ -34,11 +35,15 @@ export(renderTranslateExecuteSql)
 export(renderTranslateQueryApplyBatched)
 export(renderTranslateQuerySql)
 export(renderTranslateQuerySqlToAndromeda)
+exportClasses("Microsoft SQL Server")
```
Comment (schuemie) [inline NAMESPACE:38]: In general I prefer camel case (with 1st letter capitalized) for class names, although I guess this doesn't hurt?

**Comment by ablack3** [comment]
> I'm not quite following how this works. How does dbplyr know that for PostgresSQL, `head(5)` needs to be translated to `LIMIT 5` for PostgresSQL, and to `TOP 5` for SQL Server? Or does it rely on the fetch size of the JDBC query function?

The short answer is that

File: NAMESPACE (hunk 1 above)
Comment (ablack3) [reply NAMESPACE:38]: Yea the spaces are not great either. To get dplyr to work we'll have to use the existing connection class names (e.g. `Microsoft Sql Server`) or submit a PR to dbplyr to add a new connection class name.

**Comment by schuemie** [comment]
Ah, I see. There are existing backends for many of the SQL dialects in dbplyr itself. That is also why you're using the funky class names. That makes life a lot easier.

It seems the most important backends that are missing are BigQuery and Spark (not sure who is still using Netezza).

**Comment by ablack3** [comment]
> Ah, I see. There are existing backends for many of the SQL dialects in dbplyr itself. That is also why you're using the funky class names. That makes life a lot easier.
> 
> It seems the most important backends that are missing are BigQuery and Spark (not sure who is still using Netezza).

Right. dbplyr is providing the SQL generation and [various other driver packages](https://github.com/r-dbi/backends#readme) are providing the actual connection classes. Spark is provided by the [sparklyr package](https://spark.rstudio.com/) and BigQuery can be queried with dplyr using the [bigrquery package](https://bigrquery.r-dbi.org/). I'm not sure where the "Microsoft Sql Server" connection class is defined. Maybe the odbc package? DatabaseConnector seems similar to the RJDBC or odbc packages that provide backends for multiple databases. In addition to the DBI backend DatabaseConnector also provides more functionality. 

I'm learning a lot about the package and started building it up from scratch to get a better sense of how it works.

**Comment by ablack3** [comment]
@schuemie - Sql server dplyr translation requires dbGetInfo which should return the version of sql server in major.minor format. I wrote a naive implementation that seems to work. The function caches its value so it only queries the database once. dbGetInfo is also supposed to return the user, server, and port and should really be implemented for each connection type. dbGetInfo doesn't have access to connectionDetails. I'm not sure when we should get the database version and where we should store it.

**Comment by schuemie** [comment]
I've never called `dbGetInfo()` myself, and my current thinking is we shouldn't spend too much time on it. If dbplr needs the SQL Server version number to its work, then it is great you've now implemented that. I don't see the point in returning the user, server, and port. In fact, I would consider that sensitive information, which we handle carefully in `connectionDetails`, and shouldn't just be stored in memory without encryption. So unless dbplyr *needs* this information, I suggest we just not return it.

**Comment by ablack3** [comment]
> I've never called `dbGetInfo()` myself, and my current thinking is we shouldn't spend too much time on it. If dbplr needs the SQL Server version number to its work, then it is great you've now implemented that. I don't see the point in returning the user, server, and port. In fact, I would consider that sensitive information, which we handle carefully in `connectionDetails`, and shouldn't just be stored in memory without encryption. So unless dbplyr _needs_ this information, I suggest we just not return it.

Ok sounds good. dbplyr only needs the sql server version number so what I've done is the bare minimum to get it to work. So I think this is ready for a real review. I didn't exhaustively test dbplyr so it is possible that issues will crop up. We're also only guaranteeing it works on the databases we can test against (sql server, postgres, redshift, oracle).

**Pull request closed without merging**